statsmodels/base/model.py

```diff
@@ -510,6 +510,149 @@ def hess(params, *args):
         mlefit.mle_settings = optim_settings
         return mlefit
 
+    def _fit_zeros(self, keep_index=None, start_params=None, return_auxiliary=False,
+                   k_params=None, **fit_kwds):
+        """experimental fit the model subject to zero constraints
+
+        Intended for internal use cases until we know what we need.
+        API will need to change to handle models with two exog.
+
+        This is essentially a simplified version of `fit_constrained`, and does not need
+        to use `offset`.
+
+        The estimation creates a new model with transformed design matrix,
+        exog, and converts the results back to the original parameterization.
+
+        Some subclasses could use a more efficient calculation than using a new model.
+
+        Parameters
+        ----------
+        keep_index : array_like (int or bool) or slice
+            variables that should be dropped.
+        start_params : None or array_like
+            starting values for the optimization. `start_params` needs to be
+            given in the original parameter space and are internally
+            transformed.
+        k_params : int or None
+            If None, then we try to infer from start_params or model.
+        **fit_kwds : keyword arguments
+            fit_kwds are used in the optimization of the transformed model.
+
+        Returns
+        -------
+        results : Results instance
+
+        """
+        # wee need to append index of extra params to keep_index as in NegativeBinomial
+        if hasattr(self, 'k_extra') and self.k_extra > 0:
+            # we cannot change the original, TODO: should we add keep_index_params?
+            keep_index = np.array(keep_index, copy=True)
+            keep_index_p = np.concatenate((keep_index,
+                                           list(range(self.exog.shape[1] + self.k_extra))[-self.k_extra:]))
+        else:
+            keep_index_p = keep_index
+
+        # not all models support start_params, drop if None, hide them in fit_kwds
+        if start_params is not None:
+            fit_kwds['start_params'] = start_params[keep_index_p]
+            k_params = len(start_params) # ignore k_params in this case, or verify consisteny?
+
+        # build auxiliary model and fit
+        init_kwds = self._get_init_kwds()
+        mod_constr = self.__class__(self.endog, self.exog[:, keep_index], **init_kwds)
+        res_constr = mod_constr.fit(**fit_kwds)
+        #switch name, only need keep_index for params belos
+        keep_index = keep_index_p
+
+        if k_params is None:
+            k_params = self.exog.shape[1]
+            k_params += getattr(self, 'k_extra', 0)
+
+
+        params_full = np.zeros(k_params)
+        params_full[keep_index] = res_constr.params
+
+
+        # create dummy results Instance, TODO: wire up properly
+        # TODO: this could be moved into separate private method if needed
```

Comment:
This _very_ much should be done.  `fit_constrained` is basically copy/pasted in three places around the code-base.  This should be centralized before trying to address other issues in those methods.

Reply:
fit_zeros is simpler than fit_constrained and uses indexing instead of linear transformation.
Centralizing this makes it more difficult to make further adjustments because it needs to handle more general cases.
i.e. I still don't know how to handle #2393 in a generic and robust way.